[
  {
    "author_association": "CONTRIBUTOR",
    "body": "Personally, I advise against using `$error` for anything other than casual console debugging. \r\n\r\n> How should I monitor errors originating from a module from within that module?\r\n\r\nCould you be more specific in what you are asking? ",
    "created_at": "2018-01-02T00:25:49Z",
    "html_url": "https://github.com/PowerShell/PowerShell/issues/5768#issuecomment-354687877",
    "id": 354687877,
    "issue_url": "https://api.github.com/repos/PowerShell/PowerShell/issues/5768",
    "node_id": "MDEyOklzc3VlQ29tbWVudDM1NDY4Nzg3Nw==",
    "performed_via_github_app": null,
    "reactions": {
      "+1": 0,
      "-1": 0,
      "confused": 0,
      "eyes": 0,
      "heart": 0,
      "hooray": 0,
      "laugh": 0,
      "rocket": 0,
      "total_count": 0,
      "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354687877/reactions"
    },
    "updated_at": "2018-01-02T00:25:49Z",
    "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354687877",
    "user": {
      "avatar_url": "https://avatars.githubusercontent.com/u/6509955?v=4",
      "events_url": "https://api.github.com/users/markekraus/events{/privacy}",
      "followers_url": "https://api.github.com/users/markekraus/followers",
      "following_url": "https://api.github.com/users/markekraus/following{/other_user}",
      "gists_url": "https://api.github.com/users/markekraus/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/markekraus",
      "id": 6509955,
      "login": "markekraus",
      "node_id": "MDQ6VXNlcjY1MDk5NTU=",
      "organizations_url": "https://api.github.com/users/markekraus/orgs",
      "received_events_url": "https://api.github.com/users/markekraus/received_events",
      "repos_url": "https://api.github.com/users/markekraus/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/markekraus/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/markekraus/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/markekraus"
    }
  },
  {
    "author_association": "NONE",
    "body": ">Personally, I advise against using $error for anything other than casual console debugging.\r\n\r\nCan you explain why you advise against using `$Error`?  That advice doesn't seem consistent with [this MSDN article](https://msdn.microsoft.com/en-us/library/ms714412(v=vs.85).aspx) (and [@mklement0's pragmatic summary thereof](https://stackoverflow.com/questions/9294949/when-should-i-use-write-error-vs-throw-terminating-vs-non-terminating-errors/39949027#39949027)).  Each of those recommends using non-terminating errors in certain cases.  It seems that unit testing for non-terminating errors would naturally involve `$Error`.  Is there some other method of unit testing for non-terminating errors that I should consider?",
    "created_at": "2018-01-02T01:52:02Z",
    "html_url": "https://github.com/PowerShell/PowerShell/issues/5768#issuecomment-354692461",
    "id": 354692461,
    "issue_url": "https://api.github.com/repos/PowerShell/PowerShell/issues/5768",
    "node_id": "MDEyOklzc3VlQ29tbWVudDM1NDY5MjQ2MQ==",
    "performed_via_github_app": null,
    "reactions": {
      "+1": 0,
      "-1": 0,
      "confused": 0,
      "eyes": 0,
      "heart": 0,
      "hooray": 0,
      "laugh": 0,
      "rocket": 0,
      "total_count": 0,
      "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354692461/reactions"
    },
    "updated_at": "2018-01-02T01:52:02Z",
    "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354692461",
    "user": {
      "avatar_url": "https://avatars.githubusercontent.com/u/11237922?v=4",
      "events_url": "https://api.github.com/users/alx9r/events{/privacy}",
      "followers_url": "https://api.github.com/users/alx9r/followers",
      "following_url": "https://api.github.com/users/alx9r/following{/other_user}",
      "gists_url": "https://api.github.com/users/alx9r/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/alx9r",
      "id": 11237922,
      "login": "alx9r",
      "node_id": "MDQ6VXNlcjExMjM3OTIy",
      "organizations_url": "https://api.github.com/users/alx9r/orgs",
      "received_events_url": "https://api.github.com/users/alx9r/received_events",
      "repos_url": "https://api.github.com/users/alx9r/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/alx9r/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/alx9r/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/alx9r"
    }
  },
  {
    "author_association": "CONTRIBUTOR",
    "body": "`$Error` is populated by more than just the immediate code (nested code can populate it) and some misguided nested code will even clear it. Also, there are the scoping issues you mentioned with regards to modules. All in all, I just find `$Error` to not be very reliable. It's a handy tool when you are in the console, but I would never personally write code that depended on it or tests that run against it.\r\n\r\nFor testing non-termination errors, in this project as well as others usually involves toggling the error action preference (either the automatic variable or the via advanced cmdlet parameter) and then passing the script block to `Should Not Throw` and/or `Should Throw`. Another way is to make use of the `-ErrorVariable` parameter.",
    "created_at": "2018-01-02T10:52:34Z",
    "html_url": "https://github.com/PowerShell/PowerShell/issues/5768#issuecomment-354741922",
    "id": 354741922,
    "issue_url": "https://api.github.com/repos/PowerShell/PowerShell/issues/5768",
    "node_id": "MDEyOklzc3VlQ29tbWVudDM1NDc0MTkyMg==",
    "performed_via_github_app": null,
    "reactions": {
      "+1": 2,
      "-1": 0,
      "confused": 0,
      "eyes": 0,
      "heart": 0,
      "hooray": 0,
      "laugh": 0,
      "rocket": 0,
      "total_count": 2,
      "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354741922/reactions"
    },
    "updated_at": "2018-01-02T10:52:34Z",
    "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354741922",
    "user": {
      "avatar_url": "https://avatars.githubusercontent.com/u/6509955?v=4",
      "events_url": "https://api.github.com/users/markekraus/events{/privacy}",
      "followers_url": "https://api.github.com/users/markekraus/followers",
      "following_url": "https://api.github.com/users/markekraus/following{/other_user}",
      "gists_url": "https://api.github.com/users/markekraus/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/markekraus",
      "id": 6509955,
      "login": "markekraus",
      "node_id": "MDQ6VXNlcjY1MDk5NTU=",
      "organizations_url": "https://api.github.com/users/markekraus/orgs",
      "received_events_url": "https://api.github.com/users/markekraus/received_events",
      "repos_url": "https://api.github.com/users/markekraus/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/markekraus/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/markekraus/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/markekraus"
    }
  },
  {
    "author_association": "NONE",
    "body": ">For testing non-termination errors, in this project as well as others usually involves toggling the error action preference...\r\n\r\nThat seems pragmatic as a first approximation.  Converting non-terminating to terminating errors, however, significantly affects control flow especially where the pipeline is involved.  The case where control flow and non-terminating errors both matter is probably an edge case for most commands, but I'm finding that that's an edge case I need to test.  That's why I'm interested in `$Error`.",
    "created_at": "2018-01-02T16:26:35Z",
    "html_url": "https://github.com/PowerShell/PowerShell/issues/5768#issuecomment-354806463",
    "id": 354806463,
    "issue_url": "https://api.github.com/repos/PowerShell/PowerShell/issues/5768",
    "node_id": "MDEyOklzc3VlQ29tbWVudDM1NDgwNjQ2Mw==",
    "performed_via_github_app": null,
    "reactions": {
      "+1": 0,
      "-1": 0,
      "confused": 0,
      "eyes": 0,
      "heart": 0,
      "hooray": 0,
      "laugh": 0,
      "rocket": 0,
      "total_count": 0,
      "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354806463/reactions"
    },
    "updated_at": "2018-01-02T16:26:35Z",
    "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/354806463",
    "user": {
      "avatar_url": "https://avatars.githubusercontent.com/u/11237922?v=4",
      "events_url": "https://api.github.com/users/alx9r/events{/privacy}",
      "followers_url": "https://api.github.com/users/alx9r/followers",
      "following_url": "https://api.github.com/users/alx9r/following{/other_user}",
      "gists_url": "https://api.github.com/users/alx9r/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/alx9r",
      "id": 11237922,
      "login": "alx9r",
      "node_id": "MDQ6VXNlcjExMjM3OTIy",
      "organizations_url": "https://api.github.com/users/alx9r/orgs",
      "received_events_url": "https://api.github.com/users/alx9r/received_events",
      "repos_url": "https://api.github.com/users/alx9r/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/alx9r/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/alx9r/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/alx9r"
    }
  },
  {
    "author_association": "CONTRIBUTOR",
    "body": "@alx9r There is also `$?` which I had to use for special cases in the past to check for non-terminating errors (a different approach would be change the `ErrorAction` to `Stop` inside a try block and then access the error using `$_.Exception` in the catch block). Maybe that could help you but not sure.",
    "created_at": "2018-01-04T07:18:36Z",
    "html_url": "https://github.com/PowerShell/PowerShell/issues/5768#issuecomment-355214551",
    "id": 355214551,
    "issue_url": "https://api.github.com/repos/PowerShell/PowerShell/issues/5768",
    "node_id": "MDEyOklzc3VlQ29tbWVudDM1NTIxNDU1MQ==",
    "performed_via_github_app": null,
    "reactions": {
      "+1": 0,
      "-1": 0,
      "confused": 0,
      "eyes": 0,
      "heart": 0,
      "hooray": 0,
      "laugh": 0,
      "rocket": 0,
      "total_count": 0,
      "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/355214551/reactions"
    },
    "updated_at": "2018-01-04T07:18:36Z",
    "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/355214551",
    "user": {
      "avatar_url": "https://avatars.githubusercontent.com/u/9250262?v=4",
      "events_url": "https://api.github.com/users/bergmeister/events{/privacy}",
      "followers_url": "https://api.github.com/users/bergmeister/followers",
      "following_url": "https://api.github.com/users/bergmeister/following{/other_user}",
      "gists_url": "https://api.github.com/users/bergmeister/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/bergmeister",
      "id": 9250262,
      "login": "bergmeister",
      "node_id": "MDQ6VXNlcjkyNTAyNjI=",
      "organizations_url": "https://api.github.com/users/bergmeister/orgs",
      "received_events_url": "https://api.github.com/users/bergmeister/received_events",
      "repos_url": "https://api.github.com/users/bergmeister/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/bergmeister/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/bergmeister/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/bergmeister"
    }
  },
  {
    "author_association": "NONE",
    "body": "I have found some success monitoring the `$global:Error` for unit testing for non-terminating errors.  The scope rules of the non-global `$error`, on the other hand, remain inscrutable to me.",
    "created_at": "2018-02-03T18:16:00Z",
    "html_url": "https://github.com/PowerShell/PowerShell/issues/5768#issuecomment-362841818",
    "id": 362841818,
    "issue_url": "https://api.github.com/repos/PowerShell/PowerShell/issues/5768",
    "node_id": "MDEyOklzc3VlQ29tbWVudDM2Mjg0MTgxOA==",
    "performed_via_github_app": null,
    "reactions": {
      "+1": 0,
      "-1": 0,
      "confused": 0,
      "eyes": 0,
      "heart": 0,
      "hooray": 0,
      "laugh": 0,
      "rocket": 0,
      "total_count": 0,
      "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/362841818/reactions"
    },
    "updated_at": "2018-02-03T18:16:00Z",
    "url": "https://api.github.com/repos/PowerShell/PowerShell/issues/comments/362841818",
    "user": {
      "avatar_url": "https://avatars.githubusercontent.com/u/11237922?v=4",
      "events_url": "https://api.github.com/users/alx9r/events{/privacy}",
      "followers_url": "https://api.github.com/users/alx9r/followers",
      "following_url": "https://api.github.com/users/alx9r/following{/other_user}",
      "gists_url": "https://api.github.com/users/alx9r/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/alx9r",
      "id": 11237922,
      "login": "alx9r",
      "node_id": "MDQ6VXNlcjExMjM3OTIy",
      "organizations_url": "https://api.github.com/users/alx9r/orgs",
      "received_events_url": "https://api.github.com/users/alx9r/received_events",
      "repos_url": "https://api.github.com/users/alx9r/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/alx9r/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/alx9r/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/alx9r"
    }
  }
]
